### ISSUE TITLE
Working with experimental components

### ISSUE THREAD

**Issue by i-like-robots** [Financial-Times/x-dash#147]
We need to talk about new and experimental components. Our aims with x-dash are to provide stable modules with long term planning and well understood use cases. However, we do not wish to inhibit the creation of new modules or prevent existing (but perhaps not so great) modules from being ported over when there is a reasonable case to do so.

Matt and I have so far been of the opinion that x-dash modules should be released in lock-step. Is it possible to release modules with a stable version number and include experimental or beta quality modules with that? We plan to implement a different form of governance which does not allow bumping versions continuously as we have in the past.

I have a few ideas of how we might facilitate this without changing course too much:

1. Add an `experimental` flag to the package manifest. Modules marked as experimental will not be published and not included in the deployed documentation website and storybook.

2. Keep experimental modules in the next minor version branch which will always have several beta releases before being marked as stable.

3. Allow modules to be developed in a separate repository but enable them to "hook in" to storybook and the documentation site. They may then follow their own prerelease cycle until reaching stability from which point they can be merged into x-dash.

What ideas do you have?

**Comment by benbarnett** [comment]
> of the opinion that x-dash modules should be released in lock-step
> [..]
> We are not going to allow bumping versions continuously as we have in the past.

Can you elaborate on this? Are you picturing modules within x-dash to be released in parallel with each other?

**Comment by i-like-robots** [comment]
I've attempted to better word my original statement but yes, from the start we agreed we would like to try and follow a scheduled release pattern releasing everything under the same version number.

**Comment by kavanagh** [comment]
> Add an `experimental` flag to the package manifest. Modules marked as experimental will not be published and not included in the deployed documentation website and storybook.

And what about experimental features of an existing component? I think documenting the preferred approach for developing (and integrating into an app) would be helpful.

**Comment by kavanagh** [comment]
Allow modules to be developed in a separate repository but enable them to "hook in" to storybook and the documentation site.

I'd prefer not to do this, and stay in the same space. I think it would get messy quickly.

**Comment by benbarnett** [comment]
👍 seems like more release fanfare = better communication, and rolled up/scheduled releases will surely help with that. 

> Allow modules to be developed in a separate repository but enable them to "hook in" to storybook and the documentation site.

Although I agree with @kavanagh's comment about things potentially getting messy... for larger components/sets of components (I am thinking some of the myFT things here) – having a team's own review process could be beneficial and avoid huge/long lived branches in x-dash.

**Comment by i-like-robots** [comment]
One more small point about releasing with the same version number... this is partly influenced by the tools we use. We release on CI without ever committing version numbers to the repository. Tools which manage multi package repositories and allow individual package versions require this info. If we do find the proposed process too restrictive then we'd need to create tooling and enforce a strict naming convention or change the way we release things!

**Comment by i-like-robots** [comment]
So it looks like option 3 is off the table (I think this would actually work already but we can file it under "possible but not encouraged").

How about option 2?

> Keep experimental modules in the next minor version branch which should always have several beta releases before being marked as stable.

**Comment by kavanagh** [comment]
I think I prefer option 1 because it simplifies the development process.

So merge experimental components and ignore them from stable builds. Perhaps to give people a way to use the experimental stuff we could have CI also build all experimental stuff in an alpha build of the next major. eg 2.0.0-alpha+3fe18d `${current+1}.0.0-alpha+{git-rev}`

**Comment by i-like-robots** [comment]
> having a team's own review process could be beneficial and avoid huge/long lived branches in x-dash.

Allowing access to the toolchain for x-dash compatible components seems like it would be wise. In theory it would even be possible to pull a component into storybook from inside an app ([just stick the relative path here](https://github.com/Financial-Times/x-dash/blob/master/tools/x-workbench/register-components.js)).

> CI also build all experimental stuff in an alpha build of the next major. eg 2.0.0-alpha+3fe18d ${current+1}.0.0-alpha+{git-rev}

We could certainly do that. We'll need to define a tag naming convention for this, though I am not sure how to enforce it effectively.

**Comment by i-like-robots** [comment]
Athloi now has filters so we can execute commands for individual packages. So I propose the following process to develop and release an experimental component:

- Experimental components should remain within a branch tracking master
- The component can only be released using a pre-1.0.0 version number
- To release an experimental component the branch should be tagged following the pattern `prerelease-{component name}-v{semver}`, e.g. `prerelease-x-event-promo-v0.0.1`
- Once a component has been tested on FT.com and The App and is considered relatively stable then a PR may be raised for it to be merged into the master or next version branch.

How does that sound?

**Comment by i-like-robots** [comment]
By specifying the package name in the tag I hope that we can release only the individual component rather than everything. Athloi works by switching the working directory so `athloi publish -F x-gift-article` would be equivalent to running only `cd components/x-gift-article && npm publish`.

**Comment by i-like-robots** [comment]
@kateloschinina and @ludrob how does the proposal above sound to you?

**Comment by ludrob** [comment]
I think it's a good solution, definitely worth trying.
I can't see any real downside and it would allow us to test integration of the components with minimal overhead.
